### DISCUSSION TITLE
Use MP3EndoderLAME and MP3DecoderHelix in the same project?

### DISCUSSION THREAD

**Discussion by rasmushauschild** [pschatzmann/arduino-audio-tools#458] [Q&A] [answered]
Hi!
I'm experimenting with the different MP3 encoders and decoders, but I'm hitting a roadblock:
If I install `arduino-libhelix-main` and `arduino-liblame-main` in my project and in AudioCodecs.h I define both `USE_HELIX` and `USE_LAME`, I run into a declaration conflict for all the `LogLevelLame` and `LogLevelHelix` enumerators:

```
In file included from lib/arduino-liblame-main/src/MP3EncoderLAME.h:3,
                 from lib/arduino-audio-tools-main/src/AudioCodecs/CodecMP3LAME.h:4,
                 from lib/arduino-audio-tools-main/src/AudioCodecs/AudioCodecs.h:22,
                 from lib/arduino-audio-tools-main/src/AudioTools.h:29,
                 from src/main.cpp:7:
lib/arduino-liblame-main/src/lame_log.h:26:20: error: 'Debug' conflicts with a previous declaration
 enum LogLevelLame {Debug, Info, Warning, Error};
```

Do I need to modify the enumerator names in one of the libraries, or am I missing something?

Thanks in advance!
Rasmus

**Comment by pschatzmann** [comment, marked as answer]
It might be that there are some confilcts - but in any case you will not be able to use more then one library at the same time because there is just not enough memory!
Do not use USE_HELIX or USE_LAME (it is legacy code which will be removed soon) but use the corresponding includes like shown in the examples...

Last, but not least, I just committed a correction to liblame which should resolve the conflict...

**Reply by rasmushauschild** [reply]
Thanks for clarifying!
As far as I can tell, Helix is only used for decoding MP3 and LAME is only used for encoding MP3 - so how would you create a program that can do both, without including both libraries?

I can't seem to find this functionality in the examples.

Thanks for the help!

**Comment by pschatzmann** [comment]
You might have overlooked that the problem has been resolved. Just update the library!
I will not be able to resolve the issues related to the limited memory though - so it will only work with PSRAM or if you run the sketch on the desktop....

I would recommend to use some leaner codecs instead of MP3...

**Reply by rasmushauschild** [reply]
Ah, I see! Thank you.
Ideally I would like to use a more storage-efficient format than WAV - it's for a voice recording sketch, so audio quality and file sizes are both important factors.
Do you have a recommendation as to which codec might be best suited for that kind of job?
Perhaps the fdk-aac encoder/decoder is more efficient than MP3 in terms of RAM usage?

Thanks!

**Comment by pschatzmann** [comment]
Since it is only voice, the field is pretty much open and you can look at [all of the available options](https://github.com/pschatzmann/arduino-audio-tools/wiki/Encoding-and-Decoding-of-Audio). I suggest to have a closer look at [opus](https://opus-codec.org/) or at the low complexity codecs that are also used by Bluetooth (aptx, lc3)

**Reply by rasmushauschild** [reply]
Perfect! I will read up on those.

Thank you very much for your help!
